### PR TITLE
python37Packages.tensorflow-bin: 1.14.0 -> 2.1.0

### PR DESCRIPTION
**Pull request by cdepillabout**
###### Motivation for this change

This bumps `python37Packages.tensorflow-bin` to version 2.1.0.

It also bumps some of the dependencies of `tensorflow-bin` to the required versions:

- tensorflow-tensorboard: 1.15.0 -> 2.1.0
- tensorflow-estimator: 1.15.1 -> 2.1.0

With this PR, `tensorflow-bin` builds correctly, and appears to be able to be used correctly with `python37Packages`.

I've tested it with the following simple python script (from https://www.tensorflow.org/tutorials/quickstart/beginner):

```python
from __future__ import absolute_import, division, print_function, unicode_literals
import tensorflow as tf

mnist = tf.keras.datasets.mnist
(x_train, y_train), (x_test, y_test) = mnist.load_data()
x_train, x_test = x_train / 255.0, x_test / 255.0

model = tf.keras.models.Sequential([
  tf.keras.layers.Flatten(input_shape=(28, 28)),
  tf.keras.layers.Dense(128, activation='relu'),
  tf.keras.layers.Dropout(0.2),
  tf.keras.layers.Dense(10)
])
loss_fn = tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True)
model.compile(optimizer='adam', loss=loss_fn, metrics=['accuracy'])
model.fit(x_train, y_train, epochs=5)
model.evaluate(x_test,  y_test, verbose=2)
```

I've only tested this on Linux without CUDA.  I would appreciate if someone could try this out on Mac or with CUDA enabled.

I'm a beginner with both Tensorflow and Python, so please let me know if I am doing anything weird here.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/python-modules/tensorflow/prefetcher.sh [hunk 1]
```diff
@@ -1,33 +1,44 @@
...
+  "linux_py_35_cpu https://storage.googleapis.com/tensorflow/linux/cpu/tensorflow_cpu-${version}-cp35-cp35m-manylinux2010_x86_64.whl"
+  "linux_py_36_gpu https://storage.googleapis.com/tensorflow/linux/gpu/tensorflow_gpu-${version}-cp36-cp36m-manylinux2010_x86_64.whl"
+  "linux_py_36_cpu https://storage.googleapis.com/tensorflow/linux/cpu/tensorflow_cpu-${version}-cp36-cp36m-manylinux2010_x86_64.whl"
+  "linux_py_37_gpu https://storage.googleapis.com/tensorflow/linux/gpu/tensorflow_gpu-${version}-cp37-cp37m-manylinux2010_x86_64.whl"
+  "linux_py_37_cpu https://storage.googleapis.com/tensorflow/linux/cpu/tensorflow_cpu-${version}-cp37-cp37m-manylinux2010_x86_64.whl"
+  "mac_py_27_cpu https://storage.googleapis.com/tensorflow/mac/cpu/tensorflow-${version}-cp27-cp27m-macosx_10_9_x86_64.whl"
+  "mac_py_35_cpu https://storage.googleapis.com/tensorflow/mac/cpu/tensorflow-${version}-cp35-cp35m-macosx_10_6_intel.whl"
+  "mac_py_36_cpu https://storage.googleapis.com/tensorflow/mac/cpu/tensorflow-${version}-cp36-cp36m-macosx_10_9_x86_64.whl"
+  "mac_py_37_cpu https://storage.googleapis.com/tensorflow/mac/cpu/tensorflow-${version}-cp37-cp37m-macosx_10_9_x86_64.whl"
+)
```
Comment (cdepillabout) [inline pkgs/development/python-modules/tensorflow/prefetcher.sh]: I've tried to clean up the logic in this updater script so that it can be easier to edit in the future.

It should be pretty simple right now, so hopefully this will work for us going into the future.

File: pkgs/top-level/python-packages.nix [hunk 2]
```diff
@@ -4876,7 +4876,7 @@ in {
   protobuf = callPackage ../development/python-modules/protobuf {
     disabled = isPyPy;
     doCheck = !isPy3k;
-    protobuf = pkgs.protobuf;
+    protobuf = pkgs.protobuf3_8;
```
Comment (cdepillabout) [inline pkgs/top-level/python-packages.nix]: This is the hackiest thing in this PR.

`tensorflow-bin` requires `pythonPackages.protobuf-3.8` (and currently `pythonPackages.protobuf` is 3.7).

This will of course bump the protobuf version for all the other packages that depend on this, so it might be a big change.

I don't know if this is a reasonable thing to do (although my guess is not...).

Please let me know if there is a better way to deal with this.

**Comment by cdepillabout** [comment]
@GrahamcOfBorg build python37Packages.tensorflow-bin

**Comment by cdepillabout** [comment]
Here are two related PRs:

- https://github.com/NixOS/nixpkgs/pull/71890: bumps tensorflow-tensorboard to 2.0.0, but it should be superceded by this current PR.

- https://github.com/NixOS/nixpkgs/pull/70910: bumps tensorflow to 2.0.0.  This isn't _directly_ related to this PR, but people who are following that will probably be interested in this PR as well.  cc @tbenst @danieldk @mjlbach @Zhen-hao @Ericson2314

**Comment by cdepillabout** [comment]
I'm currently in the process of testing the following:

- Linux, non-cuda:
    - `python36Packages.tensorflow-bin`
    - `python35Packages.tensorflow-bin`
    - `python27Packages.tensorflow-bin`

(Although I'm not sure if any of them would be considered expected failures.)

**Comment by cdepillabout** [comment]
@GrahamcOfBorg build python37Packages.tensorflow-bin python36Packages.tensorflow-bin python35Packages.tensorflow-bin python27Packages.tensorflow-bin

File: pkgs/top-level/python-packages.nix (hunk 2 above)
Comment (FRidh) [reply pkgs/top-level/python-packages.nix]: The Python package protobuf is not the same as top-level protobuf. It's fine to pick another top-level protobuf, but not another Python protobuf.

**Review by FRidh — APPROVED** [review]
Did not test this but the change looks good to me.

File: pkgs/top-level/python-packages.nix (hunk 2 above)
Comment (cdepillabout) [reply pkgs/top-level/python-packages.nix]: I'm hoping that this doesn't cause some other important python package to fail build.

For instance, some Python package that requires `protobuf < 3.8`

**Comment by cdepillabout** [comment]
I forgot to update the hash for `tensorflow-tensorboard` for python27, so I just updated it and force-pushed.

File: pkgs/top-level/python-packages.nix (hunk 2 above)
Comment (FRidh) [reply pkgs/top-level/python-packages.nix]: Ah wait, I misread it initially. I thought tensorflow-bin was given another protobuf. So, we can go forward with it, but if it breaks to much it may have to be reverted.

**Comment by veprbl** [comment]
Shouldn't update to `pythonPackages.tensorflow-tensorboard` and `pythonPackages.tensorflow-estimator` break the `pythonPackages.tensorflow` unless we don't bump it as well?

File: pkgs/top-level/python-packages.nix (hunk 2 above)
Comment (cdepillabout) [reply pkgs/top-level/python-packages.nix]: > I thought tensorflow-bin was given another protobuf.

This _is_ a possibility.  I could give `tensorflow-bin` a different version of `pythonPackages.protobuf` (as well as all transitive dependencies of `tensorflow-bin` that require protobuf).

However, I imagine this might cause soundness issues?  For instance, if you try to pull in both `tensorflow-bin` and `pytorch` into the same environment, and they rely on different versions of `pythonPackages.protobuf`.

**Comment by cdepillabout** [comment]
@veprbl I was under the impression that `pythonPackages.tensorflow` was already broken.

When I try to build it on `master`, I get the following message:

```console
$ nix-build -A python37Packages.tensorflow
error: Package ‘openssl-1.0.2u’ in /home/illabout/git/nixpkgs/pkgs/development/libraries/openssl/default.nix:130 is marked as insecure, refusing to evaluate.

Known issues:
 - Support for OpenSSL 1.0.2 ended with 2019.

You can install it anyway by whitelisting this package, using the
following methods:
...
```

I don't think this should hold back updating `tensorflow-bin`.

File: pkgs/top-level/python-packages.nix (hunk 2 above)
Comment (cdepillabout) [reply pkgs/top-level/python-packages.nix]: It looks like `mysql-connector` does this:

https://github.com/NixOS/nixpkgs/blob/ddc2f887f5f4b31128e8d4a56cb524c1d36d6fd4/pkgs/development/python-modules/mysql-connector/default.nix#L4-L22

I could do a similar thing here in `tensorflow-bin`.

Comment (jonringer) [reply pkgs/top-level/python-packages.nix]: you shouldn't do this, because it will cause issues if someone creates an environment with both packages active. (Both will be present, but only 1 can be used. So one or more of the packages will be broken)

That shouldn't have been merged

**Comment by jonringer** [comment]
> @veprbl I was under the impression that `pythonPackages.tensorflow` was already broken.

This is my understanding as well, I'm fine with surpassing a broken build.

**Comment by tbenst** [comment]
tensorflow is fixed pending merge of two PRs: https://github.com/NixOS/nixpkgs/issues/80379#issuecomment-593129212.

tensorflow 2 is sadly not fully backward-compatible with tensorflow 1, and there are a number of libraries / applications that are not moving. I’d love to see both versions in nixpkgs.

**Comment by andir** [comment]
This now broke the non-bin package of tensorflow. Some of us were using it but others (like me) were trying to unbreak the edge-cases I am hitting.

TF 1.15 might have been broken in a few ways. With this PR merged we now need to unbreak even more.. I guess I'll just go about adding a second version of TF estimator for the older TF 1.15.

**Comment by FRidh** [comment]
Those that need tensorflow and related packages need to cooperate a bit more, because with all these different PR's touching it, it is hard to follow.

**Comment by timokau** [comment]
Since there seem to be many people who are very interested in keeping it working, it would be nice if some of them could add themselves as additional maintainers. Due to time constraints maintaining it isn't a very high priority for me personally right now.

**Comment by cdepillabout** [comment]
@andir Sorry for breaking the other tensorflow with this.

Could we merge in #70910 (after possibly updating it to 2.1.0)?

**Comment by Zhen-hao** [comment]
hi, what is the new `tensorflowWithCuda` for 2.1? Is it built in `tensorflow-bin_2`?

**Comment by Ericson2314** [comment]
Look at the old ones definition? All I know of is a `cudaSupport` flag on 1.x and 2.x.

**Comment by Zhen-hao** [comment]
you are right. 
```
  tensorflowWithCuda = self.tensorflow.override {
    cudaSupport = true;
  };
```
